Add a Godot 3 to Godot 4 upgrade guide (#6393)
godotengine/godot-docs: tutorials/editor/upgrading_to_godot_4.rst

.. _doc_upgrading_to_godot_4:

Upgrading to Godot 4
====================

> TODO: https://github.com/godotengine/godot-proposals/issues/387#issuecomment-1251909261

Reviewer: It's a bit weird as the first thing in the document.

If this is meant to link to further information that can be useful when doing manual porting work, I would link it as a `seealso` further down, where the user would typically need to dig deeper.

Author: That was a leftover I didn't intend to keep :slightly_smiling_face: Fixed.

    tool will *not* perform any backups of the project that is being upgraded.

    You can backup a project by using version control, or by copying the project
    folder to another location.

To use the project upgrade tool:

1. Open the Godot 4 project manager.
2. Import the Godot 3.x project using the **Import** button, or use the **Scan**
   button to find the project within a folder.
3. Double-click the imported project (or select the project then choose **Edit**).
4. You will see a dialog appearing with two options: **Convert project.godot
   Only** and **Convert Full Project**. After ensuring your project is backed up
   (see the above warning), choose **Convert Full Project**. **Convert
   project.godot Only** is intended to be used for advanced use cases *only*, in
   case the conversion tool fails.
5. Wait until the project conversion process finishes. This can take up to a few
   minutes for large projects with lots of scenes.
6. When the project manager interface becomes available again, double-click the
   project (or select the project then choose **Edit**) to open it in the
   editor.

.. note::

>     Only Godot 3.0 and later projects can be upgraded using the project
>     conversion tool found in the Godot 4 editor.

Reviewer: ```suggestion
    Only Godot 3.0 and later projects can be upgraded using the project
    conversion tool found in the Godot 4 editor.
    It is recommended to ensure that your project is up-to-date with the
    latest 3.x stable release before running the project upgrade tool.
```

    Projects made using Godot versions older than 3.0 must follow a 2-step
    process:

    1. Upgrade the project to Godot 3 using the **Tools > Export to Godot 3.0**
       option in the Godot 2.1.6 editor. Open the upgraded project in the Godot
       3.x editor, then manually fix the project to make sure it looks and
       runs correctly.

>     2. Only *after* performing step 1, you may upgrade the project to Godot 4.

Reviewer: I'm not sure this is worth describing, it literally affects maybe 2 users, who know well what they would have to do to convert to 3.x if wanted.

Fixing the project after running the project upgrade tool
---------------------------------------------------------

After upgrading the project, you may notice that certain things don't look as
they should. Scripts will likely contain various errors as well (possibly
hundreds in large projects). This is because the project upgrade tool cannot
cater to all situations. Therefore, a large part of the upgrade process remains
manual.

Automatically renamed nodes and resources
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The list below refers to the nodes' new names in Godot 4.0. In the transition
> between Godot 3.x and 4.0, dozens of nodes were renamed. 3D nodes had a ``3D``
> suffix added to them for consistency with their 2D counterparts. This conversion
> is automatic.
> 
> For ease of searching, this table lists all nodes and resources that were
> renamed in a way that does not involve only adding a ``3D`` suffix to the old name:

Reviewer: ```suggestion
The list below refers to nodes which were simply renamed for consistency or
clarity in Godot 4.0. The project upgrade tool renames them automatically in
your scripts.

One noteworthy set of renames is 3D nodes, which all got a ``3D`` suffix added for
consistency with their 2D counterparts. For example, ``Area`` is now ``Area3D``.

For ease of searching, this table lists all nodes and resources that were renamed
and are automatically converted, excluding the ones which only involved adding
a ``3D`` suffix to the old name:
```

Reviewer: I tried to make the language clearer that this is *not* all renamed nodes. IMO it's still a bit confusing though to have an "exhaustive" list that actually is only a subset.

There would need to be an intro paragraph to explain how the document is structured further (trivial renames handled by the converter, API changes which require manual handling, etc.).

But maybe there should be a table which actually lists *everything*, and has a third column that clarifies where the rename is handled by the conversion tool or not (and if not, why - different API, risk of false positives, etc.).
So users could always find an old or new name with Ctrl+F.

Author: The table would be very long and could get outdated quickly – I guess we could find a way to generate it automatically and put it on a separate page.

Reviewer 2: > But maybe there should be a table which actually lists everything, and has a third column that clarifies where the rename is handled by the conversion tool or not (and if not, why - different API, risk of false positives, etc.).
So users could always find an old or new name with Ctrl+F.

Absolutely. 100%. It's part of the reason as to why I expressed this:
> I was considering reordering that whole list 'cause it very much is... such a mess to read,

| VisualShaderNodeTextureUniform          | VisualShaderNodeTexture2DParameter        |
+-----------------------------------------+-------------------------------------------+
| VisualShaderNodeTextureUniformTriplanar | VisualShaderNodeTextureParameterTriplanar |
+-----------------------------------------+-------------------------------------------+
| VisualShaderNodeTransformUniform        | VisualShaderNodeTransformParameter        |
+-----------------------------------------+-------------------------------------------+
| VisualShaderNodeVec3Uniform             | VisualShaderNodeVec3Parameter             |
+-----------------------------------------+-------------------------------------------+
| VisualShaderNodeUniform                 | VisualShaderNodeParameter                 |
+-----------------------------------------+-------------------------------------------+
| VisualShaderNodeUniformRef              | VisualShaderNodeParameterRef              |
+-----------------------------------------+-------------------------------------------+

If you cannot find a node or resource in the list below, refer to the above

> table to find its new name.

Reviewer: The meaning of this sentence isn't really clear IMO. As a reader I don't know yet what "the list below" is, so it's confusing to be told to check the table above to find things if I can't find them...

    conversion tool found in the Godot 4 editor.

    Projects made using Godot versions older than 3.0 must follow a 2-step
    process:

    1. Upgrade the project to Godot 3 using the **Tools > Export to Godot 3.0**
       option in the Godot 2.1.6 editor. Open the upgraded project in the Godot
       3.x editor, then manually fix the project to make sure it looks and
       runs correctly.
    2. Only *after* performing step 1, you may upgrade the project to Godot 4.

Fixing the project after running the project upgrade tool
---------------------------------------------------------

After upgrading the project, you may notice that certain things don't look as
they should. Scripts will likely contain various errors as well (possibly
hundreds in large projects). This is because the project upgrade tool cannot
cater to all situations. Therefore, a large part of the upgrade process remains
manual.

Automatically renamed nodes and resources
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The list below refers to the nodes' new names in Godot 4.0. In the transition
between Godot 3.x and 4.0, dozens of nodes were renamed. 3D nodes had a ``3D``

> prefix added to them for consistency with their 2D counterparts. This conversion

Reviewer 3: ```suggestion
suffix added to them for consistency with their 2D counterparts. This conversion
```

+-----------------------------------------+-------------------------------------------+
| Physics2DDirectBodyState                | PhysicsDirectBodyState2D                  |
+-----------------------------------------+-------------------------------------------+
| Physics2DDirectSpaceState               | PhysicsDirectSpaceState2D                 |
+-----------------------------------------+-------------------------------------------+
| Physics2DServer                         | PhysicsServer2D                           |
+-----------------------------------------+-------------------------------------------+
| Physics2DShapeQueryParameters           | PhysicsShapeQueryParameters2D             |
+-----------------------------------------+-------------------------------------------+
| Physics2DTestMotionResult               | PhysicsTestMotionResult2D                 |
+-----------------------------------------+-------------------------------------------+
| PhysicsDirectBodyState                  | PhysicsDirectBodyState3D                  |
+-----------------------------------------+-------------------------------------------+

> | PhysicsDirectSpaceState                 | PhysicsDirectSpaceState3D                 |

Reviewer 3: This and DirectBodyState were only changed with a suffix being added and should be removed

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Graphics quality settings were moved from Environment properties to project
settings. This was done to make run-time quality adjustments easier, without
having to access the currently active Environment resource then modify its
properties.

As a result, you will have to configure Environment quality settings in the
project settings as old Environment quality settings aren't converted
automatically to project settings.

If you have a graphics settings menu that changed environment properties in
Godot 3.x, you will have to change its code to call :ref:`class_RenderingServer`
methods that affect environment effects' quality. Only the "base" toggle of each
environment effect and its visual knobs remain within the Environment resource.

Updating scripts to take backwards-incompatible changes into account
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some changes performed between Godot 3.x and 4 are not renames, but they still
break backwards compatibility due to different default behavior.

The most notable examples of this are:

- Built-in scripts that are :ref:`tool scripts <doc_running_code_in_the_editor>`

>   do not get the the ``tool`` keyword converted to the ``@tool`` annotation.

Reviewer 3: ```suggestion
  do not get the ``tool`` keyword converted to the ``@tool`` annotation.
```

| BakedLightmapData   | LightmapGIData        |                                                                      |
+---------------------+-----------------------+----------------------------------------------------------------------+
| BitmapFont          | FontFile              | See :ref:`doc_gui_using_fonts`.                                      |
+---------------------+-----------------------+                                                                      |
| DynamicFont         | FontFile              |                                                                      |
+---------------------+-----------------------+                                                                      |
| DynamicFontData     | FontFile              |                                                                      |
+---------------------+-----------------------+----------------------------------------------------------------------+
| Navigation3D        | Node3D                | Replaced by other Navigation nodes.                                  |
+---------------------+-----------------------+----------------------------------------------------------------------+

> | Navigation2D        | Node2D                | Replaced by other Navigation nodes.                                  |

Reviewer 3: I'd add a link to the [2D ](https://docs.godotengine.org/en/latest/tutorials/navigation/navigation_introduction_2d.html)and [3D ](https://docs.godotengine.org/en/latest/tutorials/navigation/navigation_introduction_3d.html)navigation overview pages here.
